### PR TITLE
workflow: add rune with docker and crictl integration testing

### PR DESCRIPTION
**Pull request by haosanzi**
Signed-off-by: Shirong Hao <shirong@linux.alibaba.com>

### CODE REVIEW

File: .github/workflows/rune_with_docker_and_crictl.yml [hunk 1]
```diff
@@ -0,0 +1,222 @@
...
+  # TODO: add rune circtl test on ubuntu image.
+  rune_ubuntu:
+    runs-on: self-hosted
+    steps:
+    - uses: actions/checkout@v1
+
+    - name: Get Occlum version
+      run: wget https://raw.githubusercontent.com/occlum/occlum/master/src/pal/include/occlum_version.h;
+        echo "OCCLUM_VERSION=$(grep 'Version =' occlum_version.h | awk '{print $4}')" >> $GITHUB_ENV;
+        echo "RUNE_VERSION=$(grep 'Version:' rune/dist/rpm/rune.spec | awk '{print $2}')" >> $GITHUB_ENV
```
Comment (jiazhang0) [inline .github/workflows/rune_with_docker_and_crictl.yml:16]: Use VERSION file?

File: .github/workflows/rune_with_docker_and_crictl.yml [hunk 2]
```diff
@@ -0,0 +1,222 @@
...
+
+    - name: Create container
```
Comment (jiazhang0) [inline .github/workflows/rune_with_docker_and_crictl.yml:18]: Create Occlum-$OCCLUM_VERSION for Ubuntu container

File: .github/workflows/rune_with_docker_and_crictl.yml [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+on: [push, pull_request]
+
+jobs:
+  # TODO: add rune circtl test on ubuntu image.
+  rune_ubuntu:
+    runs-on: self-hosted
+    steps:
+    - uses: actions/checkout@v1
+
+    - name: Get Occlum version
```
Comment (jiazhang0) [inline .github/workflows/rune_with_docker_and_crictl.yml:13]: Get versions

File: .github/workflows/rune_with_docker_and_crictl.yml [hunk 4]
```diff
@@ -0,0 +1,222 @@
...
+      run: wget https://raw.githubusercontent.com/occlum/occlum/master/src/pal/include/occlum_version.h;
+        echo "OCCLUM_VERSION=$(grep 'Version =' occlum_version.h | awk '{print $4}')" >> $GITHUB_ENV;
+        echo "RUNE_VERSION=$(grep 'Version:' rune/dist/rpm/rune.spec | awk '{print $2}')" >> $GITHUB_ENV
+
+    - name: Create container
+      run: docker rm -f ubuntu || true;
+        docker run -itd --privileged --rm --name=ubuntu --net host --device /dev/isgx -v /var/run/aesmd:/var/run/aesmd -v $GITHUB_WORKSPACE:/root/inclavare-containers occlum/occlum:$OCCLUM_VERSION-ubuntu18.04
```
Comment (jiazhang0) [inline .github/workflows/rune_with_docker_and_crictl.yml:20]: Use \ to split one line to multiple lines

File: .github/workflows/rune_with_docker_and_crictl.yml [hunk 5]
```diff
@@ -0,0 +1,222 @@
...
+        echo 'export GOROOT=/usr/lib/go' >> /etc/profile;
+        echo 'export GOPATH=/root/gopath' >> /etc/profile;
+        echo 'export PATH=$GOROOT/bin:$PATH:$GOPATH/bin' >> /etc/profile;
+        echo 'export GOPROXY="https://mirrors.aliyun.com/goproxy,direct"' >>/etc/profile;
+        echo 'export PKG_CONFIG_PATH=/usr/lib/pkgconfig' >>/etc/profile;
+        echo 'export GO111MODULE=on' >>/etc/profile;
+        source /etc/profile;
+        go get github.com/golang/protobuf/protoc-gen-go@v1.3.5"
+
+    - name: Build and install rune
```
Comment (jiazhang0) [inline .github/workflows/rune_with_docker_and_crictl.yml:36]: Build and install rune deb package

File: .github/workflows/rune_with_docker_and_crictl.yml [hunk 6]
```diff
@@ -0,0 +1,222 @@
...
+        cp -r inclavare-containers inclavare-containers-$RUNE_VERSION;
+        tar zcf v$RUNE_VERSION.tar.gz inclavare-containers-$RUNE_VERSION;
+        pushd /root/inclavare-containers-$RUNE_VERSION;
+        echo "$RUNE_VERSION" > VERSION;
+        sed -i '17 d' rune/dist/deb/build.sh;
+        sed -i '17icp /root/v*.tar.gz \$DEBBUILD_DIR' rune/dist/deb/build.sh;
+        source /etc/profile;
+        chmod +x rune/dist/deb/build.sh && make package components=rune;
+        dpkg -i rune_$RUNE_VERSION-1_amd64.deb;
+        popd"
```
Comment (jiazhang0) [inline .github/workflows/rune_with_docker_and_crictl.yml:47]: inclavare-containers-$RUNE_VERSION is not used any more?

File: .github/workflows/rune_with_docker_and_crictl.yml [hunk 7]
```diff
@@ -0,0 +1,222 @@
...
+              \"storage-driver\": \"vfs\"
+        }
+        EOF"
+
+        docker exec ubuntu bash -c "service docker start"
+
+    - name: Build Occlum Application Image
+      run: |
+        docker exec ubuntu bash -c "cat <<- EOF >/root/hello_world.c
+        "#include\<stdio.h\>"
```
Comment (jiazhang0) [inline .github/workflows/rune_with_docker_and_crictl.yml:74]: "#include \<stdio.h\>"

File: .github/workflows/rune_with_docker_and_crictl.yml [hunk 8]
```diff
@@ -0,0 +1,222 @@
...
+        int main ()
```
Comment (jiazhang0) [inline .github/workflows/rune_with_docker_and_crictl.yml:75]: int main (void)

File: .github/workflows/rune_with_docker_and_crictl.yml (hunk 1 above)
Comment (jiazhang0) [reply .github/workflows/rune_with_docker_and_crictl.yml:16]: occlum_version.h file should be removed?

File: .github/workflows/rune_with_docker_and_crictl.yml [hunk 9]
```diff
@@ -0,0 +1,222 @@
...
+        cd /root/inclavare-containers/rune/libenclave/internal/runtime/pal/skeleton
+        make && cp liberpal-skeleton-v*.so /usr/lib;
+        cat >Dockerfile <<-EOF
+        FROM centos:8.1.1911
+
+        RUN mkdir -p /run/rune
+        WORKDIR /run/rune
+
+        COPY encl.bin .
+        COPY encl.elf .
```
Comment (jiazhang0) [inline .github/workflows/rune_with_docker_and_crictl.yml:114]: encl.elf file is not used in runtime. Only encl.bin is necessary.

File: .github/workflows/rune_with_docker_and_crictl.yml [hunk 10]
```diff
@@ -0,0 +1,222 @@
...
+        EOF"
+
+        docker exec ubuntu bash -c "cd /root/inclavare-containers/rune/libenclave/internal/runtime/pal/skeleton;
+        docker build . -t skeleton-enclave"
+
+    - name: Rune Occlum image
+      run: docker exec ubuntu bash -c "docker run -i --rm --runtime=rune -e ENCLAVE_TYPE=intelSgx -e ENCLAVE_RUNTIME_PATH=/opt/occlum/build/lib/libocclum-pal.so -e ENCLAVE_RUNTIME_ARGS=occlum_instance occlum-app"
+
+    - name: Rune skeleton image
+      run: docker exec ubuntu bash -c "source /etc/profile; docker run -i --rm --runtime=rune -e ENCLAVE_TYPE=intelSgx -e ENCLAVE_RUNTIME_PATH=/usr/lib/liberpal-skeleton-v2.so -e ENCLAVE_RUNTIME_ARGS=debug skeleton-enclave || true"
```
Comment (jiazhang0) [inline .github/workflows/rune_with_docker_and_crictl.yml:125]: Add the test for v1 and v3 as well.

File: .github/workflows/rune_with_docker_and_crictl.yml [hunk 11]
```diff
@@ -0,0 +1,222 @@
...
+
+    - name: Kill the container
```
Comment (jiazhang0) [inline .github/workflows/rune_with_docker_and_crictl.yml:127]: Kill Occlum-$OCCLUM_VERSION for Ubuntu container

File: .github/workflows/rune_with_docker_and_crictl.yml [hunk 12]
```diff
@@ -0,0 +1,222 @@
...
+
+  rune_centos:
+    runs-on: self-hosted
+    steps:
+    - uses: actions/checkout@v1
+
+    - name: Get version
+      run: echo "RUNE_VERSION=$(grep 'Version:' rune/dist/rpm/rune.spec | awk '{print $2}')" >> $GITHUB_ENV;
+
+    - name: Create centos container
```
Comment (jiazhang0) [inline .github/workflows/rune_with_docker_and_crictl.yml:138]: Create CentOS container

**Review by jiazhang0 — CHANGES_REQUESTED** [review]

File: .github/workflows/rune_with_docker_and_crictl.yml (hunk 6 above)
Comment (haosanzi) [reply .github/workflows/rune_with_docker_and_crictl.yml:47]: yes, we use inclavare-containers-$RUNE_VERSION to create deb package.

**Review by jiazhang0 — APPROVED** [review]

**Pull request merged**